File: code/python/04-rag-pipeline/hybrid_search.py
"""Hybrid search: combine vector similarity with keyword (BM25) ranking.

Pure vector search fails when exact terms matter.  Searching for "Error 503"
may surface semantically similar documents that never mention "503".  Hybrid
search blends the ranked lists from both a vector index and a keyword index so
that exact keyword matches are rewarded without sacrificing semantic coverage.

Two fusion strategies are provided:

* **Weighted sum** — normalise each ranked list to [0, 1] then blend them
  with configurable weights (default 70 % vector / 30 % keyword).
* **Reciprocal Rank Fusion (RRF)** — position-based scoring that is robust to
  score scale differences and typically outperforms weighted sum on diverse
  queries.

Usage::

    from vector_database import SimpleVectorStore, VectorDocument
    from hybrid_search import HybridSearch

    db = SimpleVectorStore()
    # ... insert documents ...
    hs = HybridSearch(db, vector_weight=0.7)
    hs.build_keyword_index(docs)

    results = hs.search("Error 503", query_embedding=..., k=5)
    for r in results:
        print(hs.explain_result("Error 503", r))

Dependencies:
    rank_bm25 >= 0.2.2   (``pip install rank-bm25``)

See: docs/05-the-tool-ecosystem/02-vector-databases.md
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Optional

from vector_database import VectorDatabase, VectorDocument


# ---------------------------------------------------------------------------
# Data class
# ---------------------------------------------------------------------------


@dataclass
class HybridSearchResult:
    """A single result from hybrid search, carrying per-source scores."""

    id: str
    text: str
    combined_score: float
    vector_score: float
    keyword_score: float
    vector_rank: int = 0    # 1-based rank in the vector list
    keyword_rank: int = 0   # 1-based rank in the keyword list
    metadata: Optional[dict] = field(default_factory=dict)


# ---------------------------------------------------------------------------
# HybridSearch
# ---------------------------------------------------------------------------


class HybridSearch:
    """Combine vector search and BM25 keyword search.

    Args:
        vector_db:     A :class:`~vector_database.VectorDatabase` already
                       populated with documents.
        vector_weight: Weight for the vector similarity score (0–1).
                       The keyword weight is ``1 - vector_weight``.
    """

    def __init__(
        self,
        vector_db: VectorDatabase,
        vector_weight: float = 0.7,
    ) -> None:
        if not 0.0 <= vector_weight <= 1.0:
            raise ValueError("vector_weight must be in [0, 1]")
        self.vector_db = vector_db
        self.vector_weight = vector_weight
        self.keyword_weight = 1.0 - vector_weight

        # Set by build_keyword_index
        self._bm25 = None
        self._indexed_docs: list[VectorDocument] = []
        self._tokenized_corpus: list[list[str]] = []

    # ------------------------------------------------------------------
    # Index building
    # ------------------------------------------------------------------

    @staticmethod
    def _tokenize(text: str) -> list[str]:
        """Lowercase, split on non-alphanumeric, drop empty tokens."""
        return [t for t in re.split(r"[^a-z0-9]+", text.lower()) if t]

    def build_keyword_index(self, documents: list[VectorDocument]) -> None:
        """Build a BM25 index over *documents*.

        Must be called before :meth:`search` or :meth:`search_with_rrf`.

        Args:
            documents: The same documents that are stored in *vector_db*.
                       Order must be stable; the BM25 index maps positional
                       indices to these documents.
        """
        from rank_bm25 import BM25Okapi  # type: ignore[import]

        self._indexed_docs = list(documents)
        self._tokenized_corpus = [self._tokenize(doc.text) for doc in documents]
        self._bm25 = BM25Okapi(self._tokenized_corpus)

    def _require_index(self) -> None:
        if self._bm25 is None:
            raise RuntimeError("Call build_keyword_index() before searching.")

    # ------------------------------------------------------------------
    # Keyword search helpers
    # ------------------------------------------------------------------

    def _keyword_search(
        self, query: str, k: int
    ) -> list[tuple[VectorDocument, float]]:
        """Return ``(document, bm25_score)`` pairs sorted descending."""
        self._require_index()
        tokens = self._tokenize(query)
        scores = self._bm25.get_scores(tokens)  # type: ignore[union-attr]
        # Pair with documents and sort
        paired = sorted(
            zip(self._indexed_docs, scores),
            key=lambda x: x[1],
            reverse=True,
        )
        return paired[:k]

    # ------------------------------------------------------------------
    # Score normalisation
    # ------------------------------------------------------------------

    @staticmethod
    def _normalize(scores: list[float]) -> list[float]:
        """Min-max normalise *scores* to [0, 1].  Returns zeros if flat."""
        if not scores:
            return []
        lo, hi = min(scores), max(scores)
        if hi == lo:
            return [0.0] * len(scores)
        return [(s - lo) / (hi - lo) for s in scores]

    # ------------------------------------------------------------------
    # Weighted-sum hybrid search
    # ------------------------------------------------------------------

    def search(
        self,
        query: str,
        query_embedding: list[float],
        k: int = 5,
        filter_metadata: Optional[dict] = None,
    ) -> list[HybridSearchResult]:
        """Hybrid search using a weighted sum of normalised scores.

        Steps:

        1. Retrieve ``k * 2`` candidates from the vector index.
        2. Retrieve ``k * 2`` candidates from the BM25 index.
        3. Normalise both score lists to [0, 1].
        4. Combine: ``score = vector_weight * v_score + keyword_weight * k_score``.
        5. Return top *k* by combined score.

        Args:
            query:            Raw query string (used for BM25 scoring).
            query_embedding:  Embedded query vector (used for vector search).
            k:                Number of results to return.
            filter_metadata:  Optional metadata filter forwarded to the
                              vector database.

        Returns:
            List of :class:`HybridSearchResult` sorted by ``combined_score``.
        """
        self._require_index()
        fetch_k = k * 2

        # --- Vector results ---
        vector_results = self.vector_db.search(
            query_embedding, k=fetch_k, filter_metadata=filter_metadata
        )
        v_scores = self._normalize([r.score for r in vector_results])
        v_map: dict[str, tuple[float, int]] = {}  # id -> (norm_score, rank)
        for rank, (result, norm) in enumerate(zip(vector_results, v_scores), start=1):
            v_map[result.id] = (norm, rank)

        # --- Keyword results ---
        keyword_raw = self._keyword_search(query, k=fetch_k)
        k_scores = self._normalize([s for _, s in keyword_raw])
        k_map: dict[str, tuple[float, int]] = {}
        for rank, ((doc, _), norm) in enumerate(zip(keyword_raw, k_scores), start=1):
            k_map[doc.id] = (norm, rank)

        # --- Merge ---
        all_ids = {r.id for r in vector_results} | {doc.id for doc, _ in keyword_raw}
        merged: dict[str, HybridSearchResult] = {}

        for doc_id in all_ids:
            v_norm, v_rank = v_map.get(doc_id, (0.0, 0))
            k_norm, k_rank = k_map.get(doc_id, (0.0, 0))
            combined = self.vector_weight * v_norm + self.keyword_weight * k_norm

            # Look up text + metadata from vector results first, then index
            text, meta = self._lookup_text_meta(doc_id, vector_results)
            merged[doc_id] = HybridSearchResult(
                id=doc_id,
                text=text,
                combined_score=combined,
                vector_score=v_norm,
                keyword_score=k_norm,
                vector_rank=v_rank,
                keyword_rank=k_rank,
                metadata=meta,
            )

        sorted_results = sorted(
            merged.values(), key=lambda r: r.combined_score, reverse=True
        )
        return sorted_results[:k]

    # ------------------------------------------------------------------
    # Reciprocal Rank Fusion
    # ------------------------------------------------------------------

    def search_with_rrf(
        self,
        query: str,
        query_embedding: list[float],
        k: int = 5,
        rrf_k: int = 60,
        filter_metadata: Optional[dict] = None,
    ) -> list[HybridSearchResult]:
        """Hybrid search using Reciprocal Rank Fusion.

        RRF is a parameter-free fusion method that scores each document by its
        position across ranked lists::

            RRF_score(d) = Σ  1 / (rrf_k + rank_i(d))

        This is more robust than weighted-sum normalisation when the raw score
        distributions differ significantly between the two rankers.

        Args:
            query:            Raw query string.
            query_embedding:  Embedded query vector.
            k:                Number of results to return.
            rrf_k:            Ranking constant (default 60 per the original paper).
            filter_metadata:  Optional metadata filter for vector search.

        Returns:
            List of :class:`HybridSearchResult` sorted by ``combined_score``
            (which holds the RRF score here).
        """
        self._require_index()
        fetch_k = k * 2

        # --- Vector ranking ---
        vector_results = self.vector_db.search(
            query_embedding, k=fetch_k, filter_metadata=filter_metadata
        )
        v_rank: dict[str, int] = {r.id: rank for rank, r in enumerate(vector_results, start=1)}
        v_score_raw: dict[str, float] = {r.id: r.score for r in vector_results}

        # --- Keyword ranking ---
        keyword_raw = self._keyword_search(query, k=fetch_k)
        k_rank: dict[str, int] = {doc.id: rank for rank, (doc, _) in enumerate(keyword_raw, start=1)}
        k_score_raw: dict[str, float] = {doc.id: score for doc, score in keyword_raw}

        # Normalise raw scores for reporting (not used in RRF computation)
        v_norms = self._normalize(list(v_score_raw.values()))
        k_norms = self._normalize(list(k_score_raw.values()))
        v_norm_map = dict(zip(v_score_raw.keys(), v_norms))
        k_norm_map = dict(zip(k_score_raw.keys(), k_norms))

        # --- RRF fusion ---
        all_ids = set(v_rank) | set(k_rank)
        results: list[HybridSearchResult] = []

        for doc_id in all_ids:
            rrf = 0.0
            if doc_id in v_rank:
                rrf += 1.0 / (rrf_k + v_rank[doc_id])
            if doc_id in k_rank:
                rrf += 1.0 / (rrf_k + k_rank[doc_id])

            text, meta = self._lookup_text_meta(doc_id, vector_results)
            results.append(
                HybridSearchResult(
                    id=doc_id,
                    text=text,
                    combined_score=rrf,
                    vector_score=v_norm_map.get(doc_id, 0.0),
                    keyword_score=k_norm_map.get(doc_id, 0.0),
                    vector_rank=v_rank.get(doc_id, 0),
                    keyword_rank=k_rank.get(doc_id, 0),
                    metadata=meta,
                )
            )

        results.sort(key=lambda r: r.combined_score, reverse=True)
        return results[:k]

    # ------------------------------------------------------------------
    # Explanation
    # ------------------------------------------------------------------

    def explain_result(self, query: str, result: HybridSearchResult) -> str:
        """Return a human-readable explanation for a hybrid search result.

        Example output::

            [id=doc_42] combined=0.8120 | vector rank #1 (score 0.9200, weight 70%)
            + keyword rank #3 (score 0.4500, weight 30%) | query: "Error 503"

        Args:
            query:  The original query string.
            result: A :class:`HybridSearchResult` to explain.

        Returns:
            A single descriptive string.
        """
        v_part = (
            f"vector rank #{result.vector_rank} (score {result.vector_score:.4f}, "
            f"weight {self.vector_weight * 100:.0f}%)"
            if result.vector_rank
            else f"not in vector results (score 0, weight {self.vector_weight * 100:.0f}%)"
        )
        k_part = (
            f"keyword rank #{result.keyword_rank} (score {result.keyword_score:.4f}, "
            f"weight {self.keyword_weight * 100:.0f}%)"
            if result.keyword_rank
            else f"not in keyword results (score 0, weight {self.keyword_weight * 100:.0f}%)"
        )
        return (
            f"[id={result.id!r}] combined={result.combined_score:.4f} | "
            f"{v_part} + {k_part} | query: {query!r}"
        )

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _lookup_text_meta(
        self, doc_id: str, vector_results: list
    ) -> tuple[str, dict]:
        """Retrieve text and metadata for a document id.

        Prefers vector result objects (fast); falls back to the keyword index.
        """
        for r in vector_results:
            if r.id == doc_id:
                return r.text, r.metadata or {}
        for doc in self._indexed_docs:
            if doc.id == doc_id:
                return doc.text, doc.metadata or {}
        return "", {}


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------


def _run_demo() -> None:
    """Demonstrate that hybrid search ranks the exact-match document higher
    than pure vector search when the query contains a specific error code."""
    import numpy as np
    from vector_database import SimpleVectorStore

    rng = np.random.default_rng(0)
    DIM = 64

    def rand_embed() -> list[float]:
        v = rng.standard_normal(DIM)
        return (v / np.linalg.norm(v)).tolist()

    # Two very similar embeddings for "server error" documents
    base = rand_embed()
    similar = [(b + 0.05 * rng.standard_normal()) for b in base]
    similar_norm = np.array(similar)
    similar_norm = (similar_norm / np.linalg.norm(similar_norm)).tolist()

    docs = [
        VectorDocument(
            id="doc_A",
            text="Server returns error 503 when traffic exceeds limits",
            embedding=base,
        ),
        VectorDocument(
            id="doc_B",
            text="Server errors occur under heavy load",
            embedding=similar_norm,
        ),
    ]
    # Add 18 more unrelated documents
    for i in range(18):
        docs.append(
            VectorDocument(
                id=f"noise_{i}",
                text=f"Unrelated content about topic {i}",
                embedding=rand_embed(),
            )
        )

    db = SimpleVectorStore()
    db.batch_insert(docs)

    hs = HybridSearch(db, vector_weight=0.7)
    hs.build_keyword_index(docs)

    # The query embedding is similar to both doc_A and doc_B
    query_text = "Error 503"
    query_emb = base  # Maximally similar to doc_A

    print("=" * 60)
    print("Hybrid Search Demo — 'Error 503'")
    print("=" * 60)

    # --- Pure vector search ---
    print("\n[Pure vector search — top 3]")
    pure = db.search(query_emb, k=3)
    for i, r in enumerate(pure, 1):
        print(f"  #{i} id={r.id!r:12s} score={r.score:.4f}  text={r.text[:50]!r}")

    # --- Weighted hybrid search ---
    print("\n[Weighted hybrid (70% vector / 30% keyword) — top 3]")
    hybrid = hs.search(query_text, query_emb, k=3)
    for i, r in enumerate(hybrid, 1):
        print(f"  #{i} id={r.id!r:12s} combined={r.combined_score:.4f}  text={r.text[:50]!r}")
        print(f"       {hs.explain_result(query_text, r)}")

    # --- RRF hybrid search ---
    print("\n[RRF hybrid — top 3]")
    rrf = hs.search_with_rrf(query_text, query_emb, k=3)
    for i, r in enumerate(rrf, 1):
        print(f"  #{i} id={r.id!r:12s} combined={r.combined_score:.6f}  text={r.text[:50]!r}")

    # Summary: does hybrid rank doc_A first?
    doc_a_rank_pure = next((i for i, r in enumerate(pure, 1) if r.id == "doc_A"), None)
    doc_a_rank_hybrid = next((i for i, r in enumerate(hybrid, 1) if r.id == "doc_A"), None)
    doc_a_rank_rrf = next((i for i, r in enumerate(rrf, 1) if r.id == "doc_A"), None)
    print(
        f"\ndoc_A ('error 503') rank: "
        f"vector=#{doc_a_rank_pure}  "
        f"hybrid=#{doc_a_rank_hybrid}  "
        f"rrf=#{doc_a_rank_rrf}"
    )


if __name__ == "__main__":
    _run_demo()

File: code/go/04-rag-pipeline/hybrid_search.go
// hybrid_search.go — Hybrid search combining vector similarity with BM25.
//
// Pure vector search fails when exact terms matter. Hybrid search blends
// ranked lists from a vector index and a keyword (BM25) index.
//
// Two fusion strategies:
//   - Weighted sum: normalise each list to [0,1] then blend with configurable weights.
//   - Reciprocal Rank Fusion (RRF): position-based, robust to score scale differences.
//
// BM25 is implemented from scratch — no external dependencies required.
//
// See: docs/05-the-tool-ecosystem/02-vector-databases.md
package ragpipeline

import (
	"fmt"
	"math"
	"regexp"
	"sort"
	"strings"
)

// ---------------------------------------------------------------------------
// HybridSearchResult
// ---------------------------------------------------------------------------

// HybridSearchResult is a single result from hybrid search.
type HybridSearchResult struct {
	ID            string
	Text          string
	CombinedScore float64
	VectorScore   float64
	KeywordScore  float64
	VectorRank    int // 1-based rank in the vector list (0 = not in list)
	KeywordRank   int // 1-based rank in the keyword list (0 = not in list)
	Metadata      map[string]interface{}
}

// ---------------------------------------------------------------------------
// BM25 implementation
// ---------------------------------------------------------------------------

var nonAlphaNum = regexp.MustCompile(`[^a-z0-9]+`)

func tokenize(text string) []string {
	lower := strings.ToLower(text)
	tokens := nonAlphaNum.Split(lower, -1)
	var result []string
	for _, t := range tokens {
		if t != "" {
			result = append(result, t)
		}
	}
	return result
}

type bm25Index struct {
	docs      []VecDocument
	corpus    [][]string
	avgDocLen float64
	idf       map[string]float64   // IDF per term
	tf        []map[string]float64 // term freq per doc
	k1        float64
	b         float64
}

func newBM25Index(docs []VecDocument) *bm25Index {
	const k1 = 1.5
	const b = 0.75

	corpus := make([][]string, len(docs))
	tf := make([]map[string]float64, len(docs))
	totalLen := 0

	for i, doc := range docs {
		toks := tokenize(doc.Text)
		corpus[i] = toks
		totalLen += len(toks)
		freq := make(map[string]float64)
		for _, t := range toks {
			freq[t]++
		}
		tf[i] = freq
	}

	avgDocLen := 0.0
	if len(docs) > 0 {
		avgDocLen = float64(totalLen) / float64(len(docs))
	}

	// IDF: log((N - df + 0.5) / (df + 0.5) + 1)
	N := float64(len(docs))
	dfMap := make(map[string]int)
	for _, freq := range tf {
		for term := range freq {
			dfMap[term]++
		}
	}
	idf := make(map[string]float64, len(dfMap))
	for term, df := range dfMap {
		idf[term] = math.Log((N-float64(df)+0.5)/(float64(df)+0.5) + 1)
	}

	return &bm25Index{
		docs:      docs,
		corpus:    corpus,
		avgDocLen: avgDocLen,
		idf:       idf,
		tf:        tf,
		k1:        k1,
		b:         b,
	}
}

func (idx *bm25Index) score(queryTokens []string, docIdx int) float64 {
	docLen := float64(len(idx.corpus[docIdx]))
	freq := idx.tf[docIdx]
	score := 0.0
	for _, t := range queryTokens {
		idf, ok := idx.idf[t]
		if !ok {
			continue
		}
		f := freq[t]
		num := f * (idx.k1 + 1)
		denom := f + idx.k1*(1-idx.b+idx.b*docLen/idx.avgDocLen)
		score += idf * num / denom
	}
	return score
}

type bm25Hit struct {
	doc   VecDocument
	score float64
}

func (idx *bm25Index) search(query string, k int) []bm25Hit {
	tokens := tokenize(query)
	results := make([]bm25Hit, len(idx.docs))
	for i, doc := range idx.docs {
		results[i] = bm25Hit{doc: doc, score: idx.score(tokens, i)}
	}
	sort.Slice(results, func(i, j int) bool { return results[i].score > results[j].score })
	if k < len(results) {
		results = results[:k]
	}
	return results
}

// ---------------------------------------------------------------------------
// HybridSearch
// ---------------------------------------------------------------------------

// HybridSearch combines vector search and BM25 keyword search.
type HybridSearch struct {
	VectorDB      VectorDatabase
	VectorWeight  float64
	KeywordWeight float64

	keywordIdx *bm25Index
}

// NewHybridSearch creates a HybridSearch instance.
// vectorWeight is in [0,1]; keyword weight = 1 - vectorWeight.
func NewHybridSearch(vectorDB VectorDatabase, vectorWeight float64) (*HybridSearch, error) {
	if vectorWeight < 0 || vectorWeight > 1 {
		return nil, fmt.Errorf("vectorWeight must be in [0, 1]")
	}
	return &HybridSearch{
		VectorDB:      vectorDB,
		VectorWeight:  vectorWeight,
		KeywordWeight: 1 - vectorWeight,
	}, nil
}

// BuildKeywordIndex builds a BM25 index over the provided documents.
// Must be called before Search or SearchWithRRF.
func (hs *HybridSearch) BuildKeywordIndex(documents []VecDocument) {
	hs.keywordIdx = newBM25Index(documents)
}

func (hs *HybridSearch) requireIndex() error {
	if hs.keywordIdx == nil {
		return fmt.Errorf("call BuildKeywordIndex() before searching")
	}
	return nil
}

func normalizeScores(scores []float64) []float64 {
	if len(scores) == 0 {
		return nil
	}
	lo, hi := scores[0], scores[0]
	for _, s := range scores[1:] {
		if s < lo {
			lo = s
		}
		if s > hi {
			hi = s
		}
	}
	if hi == lo {
		out := make([]float64, len(scores))
		return out
	}
	out := make([]float64, len(scores))
	for i, s := range scores {
		out[i] = (s - lo) / (hi - lo)
	}
	return out
}

// Search performs hybrid search using a weighted sum of normalised scores.
func (hs *HybridSearch) Search(query string, queryEmbedding []float64, k int, filterMetadata map[string]interface{}) ([]HybridSearchResult, error) {
	if err := hs.requireIndex(); err != nil {
		return nil, err
	}
	fetchK := k * 2

	// Vector search.
	vecResults, err := hs.VectorDB.Search(queryEmbedding, fetchK, filterMetadata)
	if err != nil {
		return nil, err
	}
	vScoreRaw := make([]float64, len(vecResults))
	for i, r := range vecResults {
		vScoreRaw[i] = r.Score
	}
	vNorm := normalizeScores(vScoreRaw)
	vMap := make(map[string][2]float64) // id → {norm_score, rank}
	for rank, r := range vecResults {
		vMap[r.ID] = [2]float64{vNorm[rank], float64(rank + 1)}
	}

	// Keyword search.
	kwResults := hs.keywordIdx.search(query, fetchK)
	kScoreRaw := make([]float64, len(kwResults))
	for i, r := range kwResults {
		kScoreRaw[i] = r.score
	}
	kNorm := normalizeScores(kScoreRaw)
	kMap := make(map[string][2]float64)
	for rank, r := range kwResults {
		kMap[r.doc.ID] = [2]float64{kNorm[rank], float64(rank + 1)}
	}

	// Merge.
	allIDs := make(map[string]bool)
	for _, r := range vecResults {
		allIDs[r.ID] = true
	}
	for _, r := range kwResults {
		allIDs[r.doc.ID] = true //nolint
	}

	merged := make([]HybridSearchResult, 0, len(allIDs))
	for docID := range allIDs {
		vEntry := vMap[docID]
		kEntry := kMap[docID]
		combined := hs.VectorWeight*vEntry[0] + hs.KeywordWeight*kEntry[0]
		text, meta := hs.lookupTextMeta(docID, vecResults)
		merged = append(merged, HybridSearchResult{
			ID:            docID,
			Text:          text,
			CombinedScore: combined,
			VectorScore:   vEntry[0],
			KeywordScore:  kEntry[0],
			VectorRank:    int(vEntry[1]),
			KeywordRank:   int(kEntry[1]),
			Metadata:      meta,
		})
	}
	sort.Slice(merged, func(i, j int) bool { return merged[i].CombinedScore > merged[j].CombinedScore })
	if k < len(merged) {
		merged = merged[:k]
	}
	return merged, nil
}

// SearchWithRRF performs hybrid search using Reciprocal Rank Fusion.
func (hs *HybridSearch) SearchWithRRF(query string, queryEmbedding []float64, k, rrfK int, filterMetadata map[string]interface{}) ([]HybridSearchResult, error) {
	if err := hs.requireIndex(); err != nil {
		return nil, err
	}
	fetchK := k * 2

	vecResults, err := hs.VectorDB.Search(queryEmbedding, fetchK, filterMetadata)
	if err != nil {
		return nil, err
	}
	vRank := make(map[string]int)
	for rank, r := range vecResults {
		vRank[r.ID] = rank + 1
	}
	vScoreMap := make(map[string]float64)
	for _, r := range vecResults {
		vScoreMap[r.ID] = r.Score
	}

	kwResults := hs.keywordIdx.search(query, fetchK)
	kRank := make(map[string]int)
	for rank, r := range kwResults {
		kRank[r.doc.ID] = rank + 1
	}
	kScoreMap := make(map[string]float64)
	for _, r := range kwResults {
		kScoreMap[r.doc.ID] = r.score
	}

	// Normalize for reporting.
	vScores := make([]float64, len(vecResults))
	for i, r := range vecResults {
		vScores[i] = r.Score
	}
	kScores := make([]float64, len(kwResults))
	for i, r := range kwResults {
		kScores[i] = r.score
	}
	vNorm := normalizeScores(vScores)
	kNorm := normalizeScores(kScores)
	vNormMap := make(map[string]float64)
	for i, r := range vecResults {
		vNormMap[r.ID] = vNorm[i]
	}
	kNormMap := make(map[string]float64)
	for i, r := range kwResults {
		kNormMap[r.doc.ID] = kNorm[i]
	}

	allIDs := make(map[string]bool)
	for _, r := range vecResults {
		allIDs[r.ID] = true
	}
	for _, r := range kwResults {
		allIDs[r.doc.ID] = true
	}

	results := make([]HybridSearchResult, 0, len(allIDs))
	for docID := range allIDs {
		rrf := 0.0
		if rank, ok := vRank[docID]; ok {
			rrf += 1.0 / float64(rrfK+rank)
		}
		if rank, ok := kRank[docID]; ok {
			rrf += 1.0 / float64(rrfK+rank)
		}
		text, meta := hs.lookupTextMeta(docID, vecResults)
		results = append(results, HybridSearchResult{
			ID:            docID,
			Text:          text,
			CombinedScore: rrf,
			VectorScore:   vNormMap[docID],
			KeywordScore:  kNormMap[docID],
			VectorRank:    vRank[docID],
			KeywordRank:   kRank[docID],
			Metadata:      meta,
		})
	}
	sort.Slice(results, func(i, j int) bool { return results[i].CombinedScore > results[j].CombinedScore })
	if k < len(results) {
		results = results[:k]
	}
	return results, nil
}

// ExplainResult returns a human-readable explanation for a hybrid search result.
func (hs *HybridSearch) ExplainResult(query string, r HybridSearchResult) string {
	vPart := fmt.Sprintf("not in vector results (score 0, weight %.0f%%)", hs.VectorWeight*100)
	if r.VectorRank > 0 {
		vPart = fmt.Sprintf("vector rank #%d (score %.4f, weight %.0f%%)", r.VectorRank, r.VectorScore, hs.VectorWeight*100)
	}
	kPart := fmt.Sprintf("not in keyword results (score 0, weight %.0f%%)", hs.KeywordWeight*100)
	if r.KeywordRank > 0 {
		kPart = fmt.Sprintf("keyword rank #%d (score %.4f, weight %.0f%%)", r.KeywordRank, r.KeywordScore, hs.KeywordWeight*100)
	}
	return fmt.Sprintf("[id=%q] combined=%.4f | %s + %s | query: %q",
		r.ID, r.CombinedScore, vPart, kPart, query)
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

func (hs *HybridSearch) lookupTextMeta(docID string, vecResults []VecSearchResult) (string, map[string]interface{}) {
	for _, r := range vecResults {
		if r.ID == docID {
			return r.Text, r.Metadata
		}
	}
	if hs.keywordIdx != nil {
		for _, doc := range hs.keywordIdx.docs {
			if doc.ID == docID {
				return doc.Text, doc.Metadata
			}
		}
	}
	return "", nil
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

// RunHybridSearch demonstrates that hybrid search ranks the exact-match document
// higher than pure vector search when the query contains a specific keyword.
func RunHybridSearch() {
	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("Hybrid Search Demo — 'Error 503'")
	fmt.Println(strings.Repeat("=", 60))

	// Build a fake vector space: doc_A and doc_B have very similar embeddings,
	// but only doc_A mentions "503" in its text.
	dim := 8
	base := make([]float64, dim)
	for i := range base {
		base[i] = 1.0 / math.Sqrt(float64(dim))
	}
	similar := make([]float64, dim)
	copy(similar, base)
	similar[0] += 0.05
	norm := 0.0
	for _, v := range similar {
		norm += v * v
	}
	norm = math.Sqrt(norm)
	for i := range similar {
		similar[i] /= norm
	}

	docs := []VecDocument{
		{ID: "doc_A", Text: "Server returns error 503 when traffic exceeds limits", Embedding: base},
		{ID: "doc_B", Text: "Server errors occur under heavy load", Embedding: similar},
	}
	for i := 0; i < 8; i++ {
		v := make([]float64, dim)
		v[i%dim] = 1.0
		docs = append(docs, VecDocument{
			ID:        fmt.Sprintf("noise_%d", i),
			Text:      fmt.Sprintf("Unrelated content about topic %d", i),
			Embedding: v,
		})
	}

	db := NewInMemoryVecStore()
	db.Insert(docs)

	hs, _ := NewHybridSearch(db, 0.7)
	hs.BuildKeywordIndex(docs)

	queryText := "Error 503"
	queryEmb := base

	fmt.Println("\n[Pure vector search — top 3]")
	pure, _ := db.Search(queryEmb, 3, nil)
	for i, r := range pure {
		fmt.Printf("  #%d id=%q score=%.4f  text=%q\n", i+1, r.ID, r.Score, r.Text)
	}

	fmt.Println("\n[Weighted hybrid (70% vector / 30% keyword) — top 3]")
	hybrid, _ := hs.Search(queryText, queryEmb, 3, nil)
	for i, r := range hybrid {
		fmt.Printf("  #%d id=%q combined=%.4f  text=%q\n", i+1, r.ID, r.CombinedScore, r.Text)
		fmt.Printf("       %s\n", hs.ExplainResult(queryText, r))
	}

	fmt.Println("\n[RRF hybrid — top 3]")
	rrf, _ := hs.SearchWithRRF(queryText, queryEmb, 3, 60, nil)
	for i, r := range rrf {
		fmt.Printf("  #%d id=%q combined=%.6f  text=%q\n", i+1, r.ID, r.CombinedScore, r.Text)
	}

	findRank := func(results []HybridSearchResult, id string) int {
		for i, r := range results {
			if r.ID == id {
				return i + 1
			}
		}
		return -1
	}
	vecRankDocA := -1
	for i, r := range pure {
		if r.ID == "doc_A" {
			vecRankDocA = i + 1
			break
		}
	}
	fmt.Printf("\ndoc_A ('error 503') rank: vector=#%d  hybrid=#%d  rrf=#%d\n",
		vecRankDocA, findRank(hybrid, "doc_A"), findRank(rrf, "doc_A"))
}
